### PR TITLE
Add bidirectional RNN wrapper layer

### PR DESCRIPTION
**Pull request by AlexDBlack**
Essentially the same as Keras' Bidirectional layer.

### CODE REVIEW

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers/recurrent/Bidirectional.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+import static org.nd4j.linalg.indexing.NDArrayIndex.interval;
+import static org.nd4j.linalg.indexing.NDArrayIndex.point;
+
+/**
+ * Bidirectional is a "wrapper" layer: it wraps any uni-directional RNN layer to make it bidirectional.<br>
+ * Note that multiple different modes are supported - these specify how the activations should be combined from
+ * the forward and backward RNN networks. See {@link Mode} javadoc for more details.<br>
+ * Parameters are not shared here - there are 2 separate copies of the wrapped RNN layer, each with separate parameters.
+ * <br>
+ * Usage: {@code .layer(new Bidirecitonal(new LSTM.Builder()....build())}
```
Comment (maxpumperla) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers/recurrent/Bidirectional.java]: Bidirecitonal --> Bidirectional

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers/recurrent/Bidirectional.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+     * ADD: out = forward + backward (elementwise addition)<br>
+     * MUL: out = forward * backward (elementwise multiplication)<br>
+     * AVERAGE: out = 0.5 * (forward + backward)<br>
+     * CONCAT: Concatenate the activations.<br>
+     * Where 'forward' is the activations for the forward RNN, and 'backward' is the activations for the backward RNN.
+     * In all cases except CONCAT, the output activations size is the same size as the standard RNN that is being wrapped
+     * by this layer. In the CONCAT case, the output activations size (dimension 1) is 2x larger than the standard RNN's
+     * activations array.
+     */
+    public enum Mode {ADD, MUL, AVERAGE, CONCAT}
```
Comment (maxpumperla) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers/recurrent/Bidirectional.java]: just a remark: in keras you can also specify "NONE", which wouldn't combine the output at all, but rather return two outputs (as python list). I've personally never seen an application of this, but if it's not too difficult to achieve, might be good (esp. for keras import).

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/memory/LayerMemoryReport.java [hunk 3]
```diff
@@ -137,6 +138,31 @@ public String toString() {
         return "LayerMemoryReport(layerName=" + layerName + ",layerType=" + layerType.getSimpleName() + ")";
     }
 
+    /**
+     * Multiply all memory usage by the specified scaling factor
```
Comment (maxpumperla) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/memory/LayerMemoryReport.java]: 👍

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/recurrent/BidirectionalLayer.java [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+import static org.nd4j.linalg.indexing.NDArrayIndex.interval;
+import static org.nd4j.linalg.indexing.NDArrayIndex.point;
+
+/**
+ * Bidirectional is a "wrapper" layer: it wraps any uni-directional RNN layer to make it bidirectional.<br>
+ * Note that multiple different modes are supported - these specify how the activations should be combined from
+ * the forward and backward RNN networks. See {@link Bidirectional.Mode} javadoc for more details.<br>
+ * Parameters are not shared here - there are 2 separate copies of the wrapped RNN layer, each with separate parameters.
+ * <br>
+ * Usage: {@code .layer(new Bidirecitonal(new LSTM.Builder()....build())}
```
Comment (maxpumperla) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/recurrent/BidirectionalLayer.java]: same typo

**Comment by AlexDBlack** [comment]
@maxpumperla Should be fixed. Can't re-test DL4J yet (see gitter channel)
As for 'NONE' mode (multi-output): that'll have to wait for our big PR (that adds multi in/out support) before we can add that.

**Comment by AlexDBlack** [comment]
@maxpumperla I'll close this - both the bidirectional and simple RNN layers are here: https://github.com/deeplearning4j/deeplearning4j/pull/4398
(I rebased the simple RNN on this one as there was a bit of overlap)

**Pull request closed without merging**